ffi/derive/src/ffi_fn.rs

```diff
@@ -0,0 +1,223 @@
+        #[no_mangle]
+        #[doc = #ffi_fn_doc]
+        unsafe extern "C" #fn_signature {
+            #[allow(clippy::shadow_unrelated)]
+            let res = std::panic::catch_unwind(|| {
+                let fn_body = || #ffi_fn_body;
+
+                if let Err(err) = fn_body() {
+                    return err;
+                }
+
+                iroha_ffi::FfiReturn::Ok
+            });
+
+            match res {
+                Ok(res) => res,
+                Err(_) => {
+                    // TODO: Implement error handling (https://github.com/hyperledger/iroha/issues/2252)
+                    iroha_ffi::FfiReturn::UnrecoverableError
+                },
+            }
+        }
+    }
+}
+
+pub fn gen_fn_name(self_ty_name: Option<&Ident>, method_name: &Ident) -> Ident {
+    let self_ty_name = self_ty_name.map_or_else(Default::default, ToString::to_string);
+
+    Ident::new(
+        &format!("{}__{}", self_ty_name, method_name),
+        proc_macro2::Span::call_site(),
+    )
+}
+
+fn gen_doc(self_type: Option<&Ident>, method_name: &Ident) -> String {
+    // NOTE: [#docs = "some_doc"] expands to ///some_doc, therefore the leading space
+
+    let path = self_type.map_or_else(
+        || method_name.to_string(),
+        |self_ty| format!("{}::{}", self_ty, method_name),
+    );
+
+    format!(
+        " FFI function equivalent of [`{}`]\n \
+          \n \
+          # Safety\n \
+          \n \
+          All of the given pointers must be valid",
+        path
+    )
+}
+
+fn gen_signature(ffi_fn_name: &Ident, fn_descriptor: &FnDescriptor) -> TokenStream {
+    let self_arg = fn_descriptor
+        .receiver
+        .as_ref()
+        .map(gen_input_arg)
+        .map_or_else(Vec::new, |self_arg| vec![self_arg]);
+    let fn_args: Vec<_> = fn_descriptor.input_args.iter().map(gen_input_arg).collect();
+    let output_arg = ffi_output_arg(fn_descriptor).map(gen_out_ptr_arg);
+
+    quote! {
+        fn #ffi_fn_name<'itm>(#(#self_arg,)* #(#fn_args,)* #output_arg) -> iroha_ffi::FfiReturn
+    }
+}
+
+fn gen_input_arg(arg: &impl Arg) -> TokenStream {
+    let arg_name = arg.name();
+    let arg_type = arg.ffi_type_resolved();
+
+    quote! { #arg_name: #arg_type }
+}
+
+#[cfg(not(feature = "client"))]
+fn gen_body(fn_descriptor: &FnDescriptor) -> syn::Block {
+    let input_conversions = gen_input_conversion_stmts(fn_descriptor);
+    let method_call_stmt = gen_method_call_stmt(fn_descriptor);
+    let output_assignment = gen_output_assignment_stmts(fn_descriptor);
+
+    syn::parse_quote! {{
+        #input_conversions
+        #method_call_stmt
+        #output_assignment
+
+        Ok(())
+    }}
+}
+
+fn gen_out_ptr_arg(arg: &impl Arg) -> TokenStream {
+    let arg_name = arg.name();
+    let arg_type = arg.ffi_type_resolved();
+
+    quote! { #arg_name: <#arg_type as iroha_ffi::Output>::OutPtr }
+}
+
+#[cfg(not(feature = "client"))]
+fn gen_input_conversion_stmts(fn_descriptor: &FnDescriptor) -> TokenStream {
```

Comment:
Why do we need guards here?
Seems like it's private function.

Reply:
because it activates the lint `dead_code`